Organize into submodules

zaqqwerty opened the pull request:
Part of #99

The number of files was causing imports to become tedious.  This pull request creates two new submodules, `model` and `infer`, and moves related files into them.  New `__init__` files have been added to these modules to simplify user imports: for example, instead of
```python
from qhbmlib import circuit_infer
from qhbmlib import energy_infer
from qhbmlib import hamiltonian_infer

my_qnn= circuit_infer.QuantumInference(...)
my_ebm = energy_infer.BernoulliEnergyInference(...)
my_qhbm = hamiltonian_infer.QHBM(...)
```
we can now have
```python
from qhbmlib import infer

my_qnn = infer.QuantumInference(...)
my_ebm = infer.BernoulliEnergyInference(...)
my_qhbm = infer.QHBM(...)
```

For a followup pull request, any opinions on where the remaining top-level files should go?

farice reviewed (changes requested):
Now that we're cleaning this up, I would suggest to adopt the TFP semantic convention. We should use nouns rather than verbs for module names (sometimes plural to indicate a collection of choices). Hence, I recommend:

* `infer` -> `inference`
* `model` is OK since there are no real choices yet

Furthermore, I like the package names to emphasize key QML concepts, like inductive biases / priors. Hence:

* the losses should reside in `inference` IMO since they are just a form of MAP inference with a uniform prior.
* `priors` or `biases` or `inductive_biases` for architectures, perhaps.

I think the remainder mostly fits into something like `datasets` or `data`.

zaqqwerty commented on qhbmlib/infer/__init__.py:
```diff
@@ -0,0 +1,20 @@
+from qhbmlib.infer.ebm import EnergyInferenceBase, EnergyInference, AnalyticEnergyInference, BernoulliEnergyInference
+from qhbmlib.infer.ebm_utils import probabilities
+from qhbmlib.infer.qhbm import QHBM
+from qhbmlib.infer.qhbm_utils import density_matrix, fidelity
```
Wondering more about this, isn't use of `import *` frowned upon?  Are there use cases where wildcard imports are still the norm, which we would like to support?

farice replied:
definitely frowned upon - the real reason TF does this is because it's a form of documentation in itself (tells the programmer which objects are a part of the public API) and useful for docs (can have them auto-generated to only describe objects marked public)

zaqqwerty replied:
Makes sense - defined `__all__` in the subpackage `__init__` files

zaqqwerty commented:
Made a few changes to address your comments:

- Changed `model` to `models`, so we won't have to change the name when we add more in the future
- Changed `infer` to `inference`
- Moved `qmhl_loss.py` and `vqt_loss.py` to the `inference` subpackage
- Moved `quantum_data.py` to the new `data` subpackage

Also, decided to remove `architectures` and `hamiltonian`.  Only one function is used within the library, `get_hardware_efficient_model_unitary`, so I moved this into the `test_util` file.  Otherwise, the functions in `architectures` and `hamiltonian` were used only occasionally in research code, so if we were to keep any of it I think they only make sense in the `baselines` directory.  To this end I opened #197.

farice commented on tests/data/qhbm_data_test.py:
```diff
@@ -17,7 +17,7 @@
 import tensorflow as tf
 
-from qhbmlib import quantum_data
+from qhbmlib.data import qhbm_data
```
importing `data` doesn't work?

zaqqwerty replied:
It does work to import `data` and use `data.QHBMData` in the tests.  After you commented this, I checked some [TFP tests](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/distributions/bernoulli_test.py) and saw that they use this package import style, rather than using the explicit module name.  The style of importing the module is something I was carrying over from TFQ, [see here](https://github.com/tensorflow/quantum/blob/f10d36b64761a003fd542392767fa079d57df0aa/tensorflow_quantum/python/layers/circuit_construction/elementary_test.py#L28) for example.

My thinking was that it was better to be explicit about where code is being imported from in the library tests.  But now I think I've changed my mind: with explicit module import (like `from qhbmlib.data import qhbm_data`), we also need explicit import tests to be sure the package definitions in the `__init__` files are working, [like we do here in TFQ](https://github.com/tensorflow/quantum/blob/v0.6.1/scripts/import_test.py).  If we use the package imports in the code instead, then we know the `__init__` files are working.  When we need to test things that are not in the public API, we can fall back on using module imports.  What do you think of changing to package imports throughout the code and tests?

farice replied:
OK I see -- I think for tests explicit to the package it's ok to keep the finest granularity. But e.g. in `inference/qmhl_loss_test.py` it seems reasonable to import at `data` since that's the expected user usage pattern

farice commented on tests/inference/qmhl_loss_test.py:
```diff
@@ -22,14 +22,15 @@
-from qhbmlib import hamiltonian_infer_utils
-from qhbmlib import qmhl_loss
-from qhbmlib import quantum_data
+from qhbmlib.data import qhbm_data
```
importing `data` doesn't work?
